### PR TITLE
CompressQuantizeWeights optimizations

### PR DESCRIPTION
**Pull request by mateusztabaka**
- remove CoordinateTransform usage from FakeQuantize reference implementation
- move ZeroPointOptimizer functionality inside CompressQuantizeWeights
- compute scale and zero point in the same loop

Ticket: CVS-119273

### CODE REVIEW

**Review by jiwaszki — COMMENTED** [review]
LGTM on py side.

File: src/common/offline_transformations/include/compress_quantize_weights.hpp [hunk 1]
```diff
@@ -63,30 +62,3 @@ class ov::pass::CompressQuantizeWeights : public ov::pass::MatcherPass {
...
-                                |  (to high prec)  |
-                                +------------------+
-                                        |
-                                        v
-                  +----------+    +------------+
-                  |zero point|--->|  Subtract  |
-                  +----------+    +-----+------+
-                                        |
-                                        v
-*/
```
Comment (itikhono) [inline src/common/offline_transformations/include/compress_quantize_weights.hpp:87]: could you add this description to CompressQuantizeWeights transformation?

File: src/core/reference/include/openvino/reference/fake_quantize.hpp [hunk 2]
```diff
@@ -21,31 +21,86 @@ namespace ov {
...
     } else if (arg > std::max(in_low, in_high)) {
         return out_high;
     }
-    return static_cast<T>(std::nearbyint((arg - in_low) / (in_high - in_low) * (levels - 1)) / (levels - 1) *
+    return static_cast<T>(std::nearbyint((arg - in_low) / (in_high - in_low) * levels_minus_one) / levels_minus_one *
                               (out_high - out_low) +
                           out_low);
 }
 
+static std::vector<size_t> compute_strides(const ov::Shape& out_shape, const ov::Shape& shape);
```
Comment (itikhono) [inline src/core/reference/include/openvino/reference/fake_quantize.hpp:40]: could you describe the approach in general and what "stride", "inner_stride" and "outer_stride" mean?

File: src/common/offline_transformations/src/compress_quantize_weigths.cpp [hunk 3]
```diff
@@ -211,86 +224,630 @@ ov::pass::CompressQuantizeWeights::CompressQuantizeWeights() {
...
+                      const T* first6,
+                      U* out,
+                      F& f) {
+    while (first1 < last1) {
+        *out++ = f(*first1++, *first2++, *first3++, *first4++, *first5++, *first6++);
+    }
+}
+
+template <typename T, typename U, typename F>
+static void numpy_broadcast_4inputs(const T* weights,
```
Comment (itikhono) [inline src/common/offline_transformations/src/compress_quantize_weigths.cpp:433]: Maybe can we combine these functions and the functions implemented in fq reference and move them to some utils?

File: src/core/reference/include/openvino/reference/fake_quantize.hpp [hunk 4]
```diff
@@ -55,135 +110,396 @@ void fake_quantize(const T* const arg,
...
+                                                             const T* const input_end,
+                                                             const T* in_low,
+                                                             const T* in_high,
+                                                             const T* out_low,
+                                                             const T* out_high,
+                                                             T* out) {
+        auto in_low_scalar = *in_low;
+        auto in_high_scalar = *in_high;
+        auto out_low_scalar = *out_low;
+        auto out_high_scalar = *out_high;
```
Comment (mitruska) [inline src/core/reference/include/openvino/reference/fake_quantize.hpp:392]: Can those variables be marked as const? General comment to add missing `const` where possible to improve readability.
```suggestion
        const auto in_low_scalar = *in_low;
        const auto in_high_scalar = *in_high;
        const auto out_low_scalar = *out_low;
        const auto out_high_scalar = *out_high;
```

File: src/core/reference/include/openvino/reference/fake_quantize.hpp [hunk 5]
```diff
@@ -55,135 +110,396 @@ void fake_quantize(const T* const arg,
...
+    });
+    if (it == shape.rend()) {
+        size_t num_elements = shape_size(shape);
+        return std::tuple<size_t, size_t>{last == 1 ? 1 : num_elements,
+                                          std::min(current_output_inner_stride, num_elements)};
+    }
+    size_t idx = std::distance(it, shape.rbegin()) + static_cast<int64_t>(shape.size());
+    size_t inner_stride =
+        std::accumulate(shape.begin() + idx, shape.end(), static_cast<size_t>(1), std::multiplies<size_t>());
+    size_t weights_inner_stride = std::accumulate(output_shape.begin() + output_shape.size() - shape.size() + idx,
```
Comment (mitruska) [inline src/core/reference/include/openvino/reference/fake_quantize.hpp:273]: Consider:
```suggestion
    const size_t last = shape.back();
    auto it = std::find_if(shape.rbegin(), shape.rend(), [last](size_t dim) {
        return (last == 1 && dim > 1) || (last > 1 && dim == 1);
    });
    if (it == shape.rend()) {
        const size_t num_elements = shape_size(shape);
        return std::tuple<size_t, size_t>{last == 1 ? 1 : num_elements,
                                          std::min(current_output_inner_stride, num_elements)};
    }
    const size_t idx = std::distance(it, shape.rbegin()) + static_cast<int64_t>(shape.size());
    const size_t inner_stride =
        std::accumulate(shape.begin() + idx, shape.end(), static_cast<size_t>(1), std::multiplies<size_t>());
    const size_t weights_inner_stride = std::accumulate(output_shape.begin() + output_shape.size() - shape.size() + idx,
```

**Review by mitruska — COMMENTED** [review]
Definitely impressive work with a complex solution.
It's not easy to follow all of the changes, so some comments in the code at least for the fq reference implementation would be helpful. Also fq reference tests could be extended, if some additional cases are worth to be covered.

**Comment by itikhono** [comment]
in general, the changes look good, they increase CompressQuantizeWeights transformation performance by 20x!
@mateusztabaka please resolve the conflict with the master branch and resolve comments except this one
image
I suggest to create a ticket to resolve this comment in the next PR.

File: src/common/offline_transformations/src/compress_quantize_weigths.cpp [hunk 6]
```diff
@@ -211,86 +224,630 @@ ov::pass::CompressQuantizeWeights::CompressQuantizeWeights() {
...
-            std::all_of(adj_zero_point_val.begin(), adj_zero_point_val.end(), [](float f) -> bool {
-                return std::fabs(f) < 1e-4;
-            });
-        if (!is_adj_zero_point_close_to_zero)
+        multiply = subtract_consumers.begin()->get_node();
+    } else {
+        auto convert_to_high_precision_consumers = convert_to_high_precision->output(0).get_target_inputs();
+        if (convert_to_high_precision_consumers.size() != 1)
             return false;
+        multiply = convert_to_high_precision_consumers.begin()->get_node();
```
Comment (v-Golubev) [inline src/common/offline_transformations/src/compress_quantize_weigths.cpp:305]: Minor: why can't we use `get_single_consumer_of_type` to get multiply?

File: src/common/offline_transformations/include/compress_quantize_weights.hpp (hunk 1 above)
Comment (mateusztabaka) [reply src/common/offline_transformations/include/compress_quantize_weights.hpp:87]: done

File: src/common/offline_transformations/src/compress_quantize_weigths.cpp (hunk 3 above)
Comment (mateusztabaka) [reply src/common/offline_transformations/src/compress_quantize_weigths.cpp:425]: Maybe when we have the code that actually needs it outside of this transformation. I would prefer not to expose these symbols prematurely.

File: src/core/reference/include/openvino/reference/fake_quantize.hpp (hunk 2 above)
Comment (mateusztabaka) [reply src/core/reference/include/openvino/reference/fake_quantize.hpp:40]: done, let me know if this needs to be expanded

File: src/core/reference/include/openvino/reference/fake_quantize.hpp [hunk 7]
```diff
@@ -55,135 +110,396 @@ void fake_quantize(const T* const arg,
...
+    std::vector<size_t> output_strides = compute_strides(arg_shape, arg_shape);
+    std::vector<size_t> in_low_strides = compute_strides(arg_shape, in_low_shape);
+    std::vector<size_t> in_high_strides = compute_strides(arg_shape, in_high_shape);
+    std::vector<size_t> out_low_strides = compute_strides(arg_shape, out_low_shape);
+    std::vector<size_t> out_high_strides = compute_strides(arg_shape, out_high_shape);
 
-        while (arg3_padded_shape.size() < max_shape_size) {
-            arg3_padded_shape.insert(arg3_padded_shape.begin(), 1);
-        }
+    size_t num_elements = shape_size(arg_shape);
```
[the remark this answers is not in the document]
Comment (mateusztabaka) [reply src/core/reference/include/openvino/reference/fake_quantize.hpp]: done

File: src/core/reference/include/openvino/reference/fake_quantize.hpp [hunk 8]
```diff
@@ -55,135 +110,396 @@ void fake_quantize(const T* const arg,
...
 
-            if (arg0_padded_shape[i] == 1) {
-                arg0_squeezed_axes.insert(i);
-            } else {
-                arg0_squeezed_shape.push_back(arg0_padded_shape[i]);
-            }
+    size_t in_low_stride = 0;
+    size_t in_high_stride = 0;
+    size_t out_low_stride = 0;
+    size_t out_high_stride = 0;
```
[the remark this answers is not in the document]
Comment (mateusztabaka) [reply src/core/reference/include/openvino/reference/fake_quantize.hpp]: done

File: src/common/offline_transformations/src/compress_quantize_weigths.cpp [hunk 9]
```diff
@@ -211,86 +224,630 @@ ov::pass::CompressQuantizeWeights::CompressQuantizeWeights() {
...
+                                      const ov::Tensor& scale_tensor,
+                                      bool zero_point_is_zero,
+                                      const ov::Tensor& zero_point_tensor) {
+    ov::pass::NodeRegistry node_registry;
+    auto convert = node_registry.make<ov::op::v0::Convert>(new_weights, high_precision_type);
+    ov::pass::disable_constant_folding(convert);
+    std::shared_ptr<ov::op::v1::Multiply> mul;
+    auto scale = std::make_shared<ov::op::v0::Constant>(scale_tensor);
+    if (!zero_point_is_zero) {
+        auto zero_point = std::make_shared<ov::op::v0::Constant>(zero_point_tensor);
```
[the remark this answers is not in the document]
Comment (mateusztabaka) [reply src/common/offline_transformations/src/compress_quantize_weigths.cpp]: done

File: src/core/reference/include/openvino/reference/fake_quantize.hpp (hunk 4 above)
Comment (mateusztabaka) [reply src/core/reference/include/openvino/reference/fake_quantize.hpp]: done

File: src/common/offline_transformations/src/compress_quantize_weigths.cpp (hunk 6 above)
Comment (mateusztabaka) [reply src/common/offline_transformations/src/compress_quantize_weigths.cpp]: changed the code to use get_single_consumer_of_type

File: src/core/reference/include/openvino/reference/fake_quantize.hpp [hunk 10]
```diff
@@ -55,133 +110,451 @@ void fake_quantize(const T* const arg,
...
+    /*
+     * ---------------------------------------------------
+     * Overview:
+     *  Numpy broadcasted input tensors can be partitioned into two: outer and inner part (which also defines inner
+     *  stride as a product of inner part), so N-dimensional tensors can be processed using two loops.
+     *
+     *  For example with two inputs [2, 2, 3, 4] and [1, 1, 3, 4] we can have:
+     *      input 1 with shape [2, 2, 3, 4] can be divided into outer part [2, 2] and inner part [4, 5]
+     *      with inner stride = 12 (3 * 4) input 2 with shape [1, 1, 3, 4] can be divided into outer part [1, 1]
+     *      and inner part [4, 5] with inner stride = 12 (3 * 4)
```
Comment (v-Golubev) [inline src/core/reference/include/openvino/reference/fake_quantize.hpp:134]: Maybe I missed something but looks like inner part should be `[3, 4]`
```suggestion
     *      input 1 with shape [2, 2, 3, 4] can be divided into outer part [2, 2] and inner part [3, 4]
     *      with inner stride = 12 (3 * 4) input 2 with shape [1, 1, 3, 4] can be divided into outer part [1, 1]
     *      and inner part [3, 4] with inner stride = 12 (3 * 4)
```

File: src/core/reference/include/openvino/reference/fake_quantize.hpp [hunk 11]
```diff
@@ -55,133 +110,451 @@ void fake_quantize(const T* const arg,
...
+     *  Example with 5 inputs:
+     *      input 1 shape [2, 3, 4, 5], inner_part = [2, 3, 4, 5], inner_stride = 120
+     *      input 2 shape [1, 3, 4, 5], inner_part = [3, 4, 5], inner_stride = 60
+     *      input 3 shape [2, 3, 1, 1], inner_part = [1, 1], inner_stride = 1
+     *      input 4 shape [2, 1, 1, 1], inner_part = [1, 1, 1], inner_stride = 1
+     *      input 5 shape [1, 1, 1, 1], inner_part = [1, 1, 1, 1], inner_stride = 1
+     *
+     *      output shape [2, 3, 4, 5], inner_part = [4, 5], inner_stride = 20
+     *
+     *      Inner part with fewest number of elements is [1, 1] for input 2. So the inner part for output shape is [4, 5]
```
Comment (v-Golubev) [inline src/core/reference/include/openvino/reference/fake_quantize.hpp:177]: > Inner part with fewest number of elements is [1, 1] for input 2

Input 2 has other inner part:
`input 2 shape [1, 3, 4, 5], inner_part = [3, 4, 5]`.
Did you mean input 3 maybe?
`input 3 shape [2, 3, 1, 1], inner_part = [1, 1]`
Or we can numerate inputs starting with `0`.

File: src/common/transformations/tests/utils/compress_quantize_weights.cpp [hunk 12]
```diff
@@ -109,9 +117,10 @@ static std::vector<CompressQuantizeWeightsParams> params = {
      6.0f,
      17,
      element::i8,
-     {-8.0f, -8.0f, -8.0f, -6.0f, -4.0f, -2.0f, 0.0f, 8.0f},
+     {-4.0f, -4.0f, -4.0f, -2.0f, 0.0f, 2.0f, 4.0f, 12.0f},
```
[the remark this answers is not in the document]
Comment (v-Golubev) [reply src/common/transformations/tests/utils/compress_quantize_weights.cpp:120]: Thanks for the explanation

**Review by v-Golubev — APPROVED** [review]
In general LGTM, I have only 2 minor comments regarding description

File: src/core/reference/include/openvino/reference/fake_quantize.hpp (hunk 10 above)
Comment (mateusztabaka) [reply src/core/reference/include/openvino/reference/fake_quantize.hpp:134]: you're right. I've had different values initially - looks like I failed to change that part of the comment.

File: src/core/reference/include/openvino/reference/fake_quantize.hpp (hunk 11 above)
Comment (mateusztabaka) [reply src/core/reference/include/openvino/reference/fake_quantize.hpp]: > Did you mean input 3 maybe?

yes, changed that to 3

**Review by itikhono — APPROVED** [review]

**Review by mitruska — APPROVED** [review]

**Pull request merged**